# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/workers/ScriptLoader.cpp
# Commit: 6c129bd72b61
# Full Hash: 6c129bd72b61cf6d096dcec6c7bb188e3ddb37c7
# Author: Yulia Startsev <ystartsev@mozilla.com>
# Date: 2022-12-15 09:27:59
# Regressor Bug: 1247687
# File Overlap Count: 1
# Description:
#   Bug 1247687 - Implement InitModuleLoader methods for WorkerScriptLoader and WorkerGlobalScope; r=jonco
#   
#   This implements a method to initialize the moduleLoader for workers. This will initialize only once, for all worker types (module and classic).
#   
#   Depends on D147324
# ==============================================================================

diff -r 4b0a4fcc6894 -r 6c129bd72b61 dom/workers/ScriptLoader.cpp
--- a/dom/workers/ScriptLoader.cpp	Wed Dec 14 14:55:34 2022 +0000
+++ b/dom/workers/ScriptLoader.cpp	Wed Dec 14 14:55:35 2022 +0000
@@ -411,8 +411,20 @@
   }
 
   nsIGlobalObject* global = GetGlobal();
+  mController = global->GetController();
+  // Set up the module loader, if it has not been yet.
+  // TODO: Implement this for classic scripts when dynamic imports are added.
+  if (aWorkerPrivate->WorkerType() == WorkerType::Module) {
+    InitModuleLoader();
+  }
+}
 
-  mController = global->GetController();
+void WorkerScriptLoader::InitModuleLoader() {
+  mWorkerRef->Private()->AssertIsOnWorkerThread();
+  RefPtr<WorkerModuleLoader> moduleLoader =
+      new WorkerModuleLoader(this, GetGlobal(), mSyncLoopTarget.get());
+  static_cast<WorkerGlobalScopeBase*>(GetGlobal())
+      ->InitModuleLoader(moduleLoader);
 }
 
 void WorkerScriptLoader::CreateScriptRequests(